manuals/msiecfexport.1
.Dd June 25, 2026
.Dt MSIECFEXPORT 1
.Os
.Sh NAME
.Nm msiecfexport
.Nd exports items stored in a MSIE Cache File (index.dat)
.Sh SYNOPSIS
.Nm msiecfexport
.Op Fl c Ar codepage
.Op Fl l Ar log_file
.Op Fl m Ar mode
.Op Fl t Ar target
.Op Fl hvV
.Ar source
.Sh DESCRIPTION
.Nm msiecfexport
is a utility to export items stored in a MSIE Cache File (index.dat)
.Pp
.Nm msiecfexport
is part of the
.Nm libmsiecf
package.
.Nm libmsiecf
is a library to access the MSIE Cache File (index.dat) format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c Ar codepage
codepage of ASCII strings, options: ascii, windows-874, windows-932, \
windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 \
(default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 \
or windows-1258
.It Fl h
shows this help
.It Fl l Ar log_file
logs information about the exported items
.It Fl m Ar mode
specify the export mode, option: all, items (default), recovered 'all' exports \
the (allocated) items and recovered items, 'items' exports the (allocated) \
items and 'recovered' exports
the recovered items
.It Fl t Ar target
specify the target directory to export to (default is the source filename \
followed by .export)
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# msiecfexport index.dat
msiecfexport 20260625
.sp
Opening file.
Exporting items.
Number of items: 1.
.sp
Record type		: URL
Offset range		: 20480 - 20736 (256)
Location		: Cookie:username@domain.com/
Primary time		: Mar 23, 2009 17:23:01.515000
Secondary time		: Nov 29, 2007 11:23:44.906000
Filename		: username@domain[1].txt
Cache directory index	: -2 (0xfe)
.sp
Exporting recovered items.
Number of recovered items: 0.
.sp
Export completed.
.sp
Export completed.
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr msiecfinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libmsiecf/issues
.Sh COPYRIGHT
Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* address clean up TODOs
* complete thread lock support
** leak
** redirected
** url

* pymsiecf
  - add support for leak: Last cache synchronization date and time
* hash calculation: check what the preceding bytes could be
* use libfdata ?
* msiecfexport:
  - URL what about unknown timestamps
* code clean up
  - add initialize functions
  - rename item values to item value
  - change read values to match libwrc

Tests:
* add cache directory tests

File format
* implement MSIE 3 cache file formats ?
* access records not referred by hash table
  - flag if items are hashed

Performance improvement
* pre allocate item/directory array entries

Tools
* msiecfinfo
  - print cache directories
* msiecfexport
  - export to TSV ?
  - flexible output formatting
  - flexible date/time string formatting

In progress:
* record scan
  - flag entries in hash table as hashed

20140317
* see `git log' for more recent change log
* worked on setup.py

20140131
* 2014 update
* updated dependencies

20131015
* updated dependencies
* updated msvscpp files

20131008
* improved pymsiecf as integer functions

20130929
* worked on setup.py, largely for MSI builds

20120923
* updated msvscpp files

20120922
* updated dependencies

20120806
* fixes for building with stand-alone libuna and libbfio

20120805
* updated dependencies
* bug fix in pymsiecf file object libbfio glue code
* worked on tests

20120529
* updated dependencies

20120528
* worked on pymsiecf simple inheritance of leak, redirected and url items

20120425
* worked on hash calculation

20120423
* added resource file to msvscpp files

20120421
* worked on tests

20120417
* bug fix in record scan due to recent dependency update

20120416
* code clean up

20120414
* textual changes
* worked on tests
* updates and bug fixes in pymsiecf
* fixed codepage 1255 restriction

20120413
* updated dependencies

20120409
* fixed codepage 949 and 950 restriction

20120319
* updated dependencies

20130317
* Small updates

20130310
* worked on url item type support
* msiecfinfo: small changes to output, added major and minor version
* worked on MSIECF 4.7 format support

20130306
* worked on libfole support

20130305
* changes for pymsiecf
* updated libfdatetime
* msiecfexport added expiration and last checked time

20130304
* updated PackageMaker files
* bug fix in pymsiecf FAT date time conversion

20130119
* updated rpm files to include Python bindings

20130114
* updated dpkg files to include Python bindings
* pymsiecf: added generic item fallback object

20130113
* worked on msvscpp files

20130112
* worked on pymsiecf
* fixed issue in spec files

20130111
* worked on pymsiecf

20130110
* removed deprecated functions
* added missing API functions
* worked on pymsiecf

20130109
* 2013 update
* updated dependencies

20120930
* updated dependencies
* changes for libcdata update

20120627
* updated dependencies
* reversed library order in tools for mingw

20120428
* updated dependencies

20120102
* 2012 update

20111120
* updated libfvalue

20111114
* code clean up
* updated array type, list type, offset list
* updated libcstring, libsystem, libfdatetime, libfvalue

20111101
* fix in msiecftools to set codepage

20111031
* updated libcstring, libsystem, libuna, libbfio, libfdatetime, libfvalue

20111030
* worked on UTF8 and UTF16 support
* code clean up
  - added info handle
  - added export handle
* updated msvscpp files

20111029
* updated m4 files
* updated libfdatetime
* added libfvalue
* code clean up
  - removed libmsiecf_item_attach
  - removed libmsiecf_item_detach
* worked on UTF8 and UTF16 support

20111009
* 2011 update
* updated configure.ac and m4 files
* updated include/codepage.h, include/types.h, include/error.h
* updated common
* updated array type, list type, offset list
* updated libcstring, liberror, libnotify, libsystem, libuna, libbfio,
  libfdatetime
* updated pc and spec file
* added gettext support
* updates msvscpp files
* added unused macros
* code clean up
* fixed some leaks in error paths

20101111
* updated configure.ac
* updated include/error.h, include/types.h
* updated common
* updated array type, offset list
* updated liberror, libsystem
* added libmsiecf.rc
* code clean up
* updated msvscpp files

20101023
* added system string support in debug output of url values

20100917
* code clean up
* worked on libcstring support
* worked on offset list changes
* updated log handle
* fix for clean-up bug

20100916
* worked on offset list update
* updated liberror, libbfio

20100912
* email change
* updated configure
* updated include/codepage.h
* updated include/error.h
* updated include/types.h
* added libcstring
* updated common, liberror, libnotify, libsystem, libuna, libbfio,
  libfdatetime 
* updated array type

20100319
* fix in msvscpp files

20100114
* small changes and fixes
* updated check signature
* fix for date time string size for invalid date/time values

20010113
* 2010 update
* updated common, liberror, libnotify, libsystem, libfdatetime, libuna, libbfio
* small changes
* update for libnotify

20091220
* updated libuna
* worked on generic item layer

20091219
* msiecfexport: changed default codepage
* msiecfexport: now used libfdatetime
* library: worked on codepage support
* worked on man pages

20091217
* updated common, libnotify, liberror, libuna, libbfio, libfdatetime
* replaced endian by byte_stream
* worked on generic item layer
  - removed item reference list

20091013
* small adjustments

20091010
* small adjustments

20091009
* worked on msvscpp project files
* updated libmsiecf_notify

20091008
* worked on msvscpp project files
* updated common, libuna, libbfio, liberror, libnotify, libsystem
* updated list and array type

20090912
* worked on file format

20090911
* updated libfdatetime, libbfio, libuna, liberror, libnotify, common
* added libsystem
* added support for m4 directory
* updated spec file and support for empty requires and build requires
* worked on local library support

msiecftools/msiecfexport.c
/*
 * Extracts items from from a MSIE Cache File (index.dat).
 *
 * Copyright (C) 2009-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "export_handle.h"
#include "log_handle.h"
#include "msiecftools_getopt.h"
#include "msiecftools_libcerror.h"
#include "msiecftools_libclocale.h"
#include "msiecftools_libcnotify.h"
#include "msiecftools_libmsiecf.h"
#include "msiecftools_output.h"
#include "msiecftools_signal.h"
#include "msiecftools_unused.h"

export_handle_t *msiecfexport_export_handle = NULL;
int msiecfexport_abort                      = 0;

/* Signal handler for msiecfexport
 */
void msiecfexport_signal_handler(
      msiecftools_signal_t signal MSIECFTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "msiecfexport_signal_handler";

	MSIECFTOOLS_UNREFERENCED_PARAMETER( signal )

	msiecfexport_abort = 1;

	if( msiecfexport_export_handle != NULL )
	{
		if( export_handle_signal_abort(
		     msiecfexport_export_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal export handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use msiecfexport to export items stored in from a MSIE Cache File (index.dat).";

	msiecftools_option_t options[ ] = {
		{ 'c', "codepage", "codepage of ASCII strings, options: ascii, windows-874, windows-932, windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 (default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 or windows-1258" },
		{ 'h', NULL, "shows this help" },
		{ 'l', "log_file", "logs information about the exported items" },
		{ 'l', "mode", "export mode, option: all, items (default), recovered. 'all' exports the (allocated) items and recovered items, 'items' exports the (allocated) items and 'recovered' exports the recovered items" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source file" },
	};
	system_character_t options_string[ 32 ];

	libcerror_error_t *error               = NULL;
	log_handle_t *log_handle               = NULL;
	system_character_t *log_filename       = NULL;
	system_character_t *option_codepage    = NULL;
	system_character_t *option_export_mode = NULL;
	system_character_t *source             = NULL;
	char *program                          = "msiecfexport";
	system_integer_t option                = 0;
	int number_of_options                  = (int) ( sizeof( options ) / sizeof( msiecftools_option_t ) );
	int result                             = 0;
	int verbose                            = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "msiecftools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( msiecftools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	msiecftools_output_version_fprint(
	 stdout,
	 program );

	if( msiecftools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = msiecftools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				msiecftools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'c':
				option_codepage = optarg;

				break;

			case (system_integer_t) 'h':
				msiecftools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'l':
				log_filename = optarg;

				break;

			case (system_integer_t) 'm':
				option_export_mode = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				msiecftools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		msiecftools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libmsiecf_notify_set_stream(
	 stderr,
	 NULL );
	libmsiecf_notify_set_verbose(
	 verbose );

	if( log_handle_initialize(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize log handle.\n" );

		goto on_error;
	}
	if( export_handle_initialize(
	     &msiecfexport_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize export handle.\n" );

		goto on_error;
	}
	if( option_export_mode != NULL )
	{
		result = export_handle_set_export_mode(
			  msiecfexport_export_handle,
			  option_export_mode,
			  &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set export mode.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported export mode defaulting to: items.\n" );
		}
	}
	if( option_codepage != NULL )
	{
		result = export_handle_set_ascii_codepage(
		          msiecfexport_export_handle,
		          option_codepage,
		          &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set ASCII codepage in export handle.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported ASCII codepage defaulting to: windows-1252.\n" );
		}
	}
	if( log_handle_open(
	     log_handle,
	     log_filename,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open log file: %" PRIs_SYSTEM ".\n",
		 log_filename );

		goto on_error;
	}
	if( export_handle_open(
	     msiecfexport_export_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	result = export_handle_export_file(
	          msiecfexport_export_handle,
	          log_handle,
	          &error );

	if( result == -1 )
	{
		fprintf(
		 stderr,
		 "Unable to export file.\n" );

		goto on_error;
	}
	if( export_handle_close(
	     msiecfexport_export_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close export handle.\n" );

		goto on_error;
	}
	if( export_handle_free(
	     &msiecfexport_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free export handle.\n" );

		goto on_error;
	}
	if( log_handle_close(
	     log_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close log handle.\n" );

		goto on_error;
	}
	if( log_handle_free(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free log handle.\n" );

		goto on_error;
	}
	if( result == 0 )
	{
		fprintf(
		 stdout,
		 "No aliases to export.\n" );
	}
	else
	{
		fprintf(
		 stdout,
		 "Export completed.\n" );
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( msiecfexport_export_handle != NULL )
	{
		export_handle_free(
		 &msiecfexport_export_handle,
		 NULL );
	}
	if( log_handle != NULL )
	{
		log_handle_free(
		 &log_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

